skgstat/MetricSpace.py

```diff
@@ -243,3 +243,70 @@ def dists(self):
 
         # return
         return self._dists
+
+class ProbabalisticMetricSpace(MetricSpace):
+    """Like MetricSpace but samples the distance pairs only returning a
+       `samples` sized subset. `samples` can either be a fraction of
+       the total number of pairs (float < 1), or an integer count.
+    """
+    
+    def __init__(self, coords, dist_metric="euclidean", max_dist=None, samples=0.5):
+        self.coords = coords.copy()
+        self.dist_metric = dist_metric
+        self.max_dist = max_dist
+        self.samples = samples
+
+        self._lidx = None
+        self._ridx = None
+        self._ltree = None
+        self._rtree = None
+        self._dists = None
+        # Do a very quick check to see throw exceptions if self.dist_metric is invalid...
+        scipy.spatial.distance.pdist(self.coords[:1,:], metric=self.dist_metric)
+
+    @property
+    def sample_count(self):
+        if isinstance(self.samples, int):
+            return self.samples
+        return int(self.samples * len(self.coords))
+        
+    @property
+    def lidx(self):
+        if self._lidx is None:
+            self._lidx = np.random.choice(len(self.coords), size=self.sample_count, replace=False)
+        return self._lidx
+    
+    @property
+    def ridx(self):
+        if self._ridx is None:
+            self._ridx = np.random.choice(len(self.coords), size=self.sample_count, replace=False)
+        return self._ridx
+    
+    @property
+    def ltree(self):
+        assert self.dist_metric == "euclidean", "A coordinate tree can only be constructed for an euclidean space"
+        if self._ltree is None:
+            self._ltree = scipy.spatial.cKDTree(self.coords[self.lidx,:])
+        return self._ltree
+
+    @property
+    def rtree(self):
+        assert self.dist_metric == "euclidean", "A coordinate tree can only be constructed for an euclidean space"
+        if self._rtree is None:
+            self._rtree = scipy.spatial.cKDTree(self.coords[self.ridx,:])
+        return self._rtree
+            
+    @property
+    def dists(self):
+        if self._dists is None:
+            max_dist = self.max_dist
+            if max_dist is None:
+                max_dist = 1 + np.sqrt(((self.coords.max(axis=0) - self.coords.min(axis=0))**2).sum())
```

Comment:
@redhog, is this still the intended behavior if the distance metric is not euklidean?

Reply:
Oh, good catch. This is meant to create some value that's greater than the greatest possible distance.... Maybe just set ot to float max actually...